File: app/src/vdsensor/esp3/packets.py
"""ESP3 enums — packet types, return codes, event codes, command codes.

References are to ESP3 v1.58 (file EnOceanSerialProtocol3-1.pdf at the repo root).
"""

from __future__ import annotations

from enum import IntEnum

SYNC_BYTE = 0x55                                # §3.4.1
HEADER_LENGTH = 4                               # u16DataLength + u8OptLen + u8Type


class PacketType(IntEnum):
    """ESP3 packet type field (§1.8 / per-section structures)."""

    RADIO_ERP1 = 0x01                           # §2.1
    RESPONSE = 0x02                             # §2.2
    RADIO_SUB_TEL = 0x03                        # §2.3
    EVENT = 0x04                                # §2.4
    COMMON_COMMAND = 0x05                       # §2.5
    SMART_ACK_COMMAND = 0x06                    # §2.6
    REMOTE_MAN_COMMAND = 0x07                   # §2.7
    RADIO_MESSAGE = 0x09                        # §2.8
    RADIO_ERP2 = 0x0A                           # §2.9
    COMMAND_ACCEPTED = 0x0C                     # §2.10
    RADIO_802_15_4 = 0x10                       # §2.11
    CONFIG_2_4_GHZ = 0x11                       # §2.12


class ReturnCode(IntEnum):
    """RESPONSE return codes (§2.2.2)."""

    RET_OK = 0x00
    RET_ERROR = 0x01
    RET_NOT_SUPPORTED = 0x02
    RET_WRONG_PARAM = 0x03
    RET_OPERATION_DENIED = 0x04
    RET_LOCK_SET = 0x05
    RET_BUFFER_TO_SMALL = 0x06
    RET_NO_FREE_BUFFER = 0x07


class EventCode(IntEnum):
    """EVENT (0x04) event codes (§2.4.2)."""

    SA_RECLAIM_NOT_SUCCESSFUL = 0x01            # §2.4.3
    SA_CONFIRM_LEARN = 0x02                     # §2.4.4
    SA_LEARN_ACK = 0x03                         # §2.4.5
    CO_READY = 0x04                             # §2.4.6
    CO_EVENT_SECUREDEVICES = 0x05               # §2.4.7
    CO_DUTYCYCLE_LIMIT = 0x06                   # §2.4.8
    CO_TRANSMIT_FAILED = 0x07                   # §2.4.9
    CO_TX_DONE = 0x08                           # §2.4.10
    CO_LRN_MODE_DISABLED = 0x09                 # §2.4.11


class CommonCommand(IntEnum):
    """COMMON_COMMAND (0x05) command codes (§2.5.2)."""

    CO_WR_SLEEP = 0x01                          # §2.5.3
    CO_WR_RESET = 0x02                          # §2.5.4
    CO_RD_VERSION = 0x03                        # §2.5.5
    CO_RD_SYS_LOG = 0x04                        # §2.5.6
    CO_WR_IDBASE = 0x07                         # §2.5.9
    CO_RD_IDBASE = 0x08                         # §2.5.10
    CO_WR_LEARNMODE = 0x17                      # §2.5.25
    CO_RD_LEARNMODE = 0x18                      # §2.5.26


# RORG bytes (radio choice byte) — only the ones we decode in v1
class Rorg(IntEnum):
    RPS = 0xF6                                  # §2.1 / EEP F6
    ONE_BS = 0xD5                               # §2.1 / EEP D5
    FOUR_BS = 0xA5                              # §2.1 / EEP A5
    VLD = 0xD2                                  # §2.1 / EEP D2
    UTE = 0xD4                                  # §2.1 / EEP D4
    ADT = 0xA6                                  # §2.1 (addressed destination wrapper)
    SIGNAL = 0xD0                               # §2.1
    MSC = 0xD1                                  # §2.1

File: app-rs/src/esp3/packets.rs
//! ESP3 enums + protocol constants. All numeric values reference ESP3 v1.58.

/// §3.4.1 — every ESP3 frame starts with this byte.
pub const SYNC_BYTE: u8 = 0x55;

/// §1.7 — header is u16 data length + u8 opt length + u8 packet type.
pub const HEADER_LENGTH: usize = 4;

/// §1.8 / §2.x — packet-type field discriminator.
#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum PacketType {
    RadioErp1 = 0x01,         // §2.1
    Response = 0x02,          // §2.2
    RadioSubTel = 0x03,       // §2.3
    Event = 0x04,             // §2.4
    CommonCommand = 0x05,     // §2.5
    SmartAckCommand = 0x06,   // §2.6
    RemoteManCommand = 0x07,  // §2.7
    RadioMessage = 0x09,      // §2.8
    RadioErp2 = 0x0A,         // §2.9
    CommandAccepted = 0x0C,   // §2.10
    Radio80215_4 = 0x10,      // §2.11
    Config24Ghz = 0x11,       // §2.12
}

/// §2.2.2 — RESPONSE return codes.
#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ReturnCode {
    RetOk = 0x00,
    RetError = 0x01,
    RetNotSupported = 0x02,
    RetWrongParam = 0x03,
    RetOperationDenied = 0x04,
    RetLockSet = 0x05,
    RetBufferToSmall = 0x06,
    RetNoFreeBuffer = 0x07,
}

/// §2.4.2 — EVENT (0x04) event codes.
#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum EventCode {
    SaReclaimNotSuccessful = 0x01,
    SaConfirmLearn = 0x02,
    SaLearnAck = 0x03,
    CoReady = 0x04,
    CoEventSecuredevices = 0x05,
    CoDutycycleLimit = 0x06,
    CoTransmitFailed = 0x07,
    CoTxDone = 0x08,
    CoLrnModeDisabled = 0x09,
}

/// §2.5.2 — COMMON_COMMAND (0x05) command codes.
#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum CommonCommand {
    CoWrSleep = 0x01,
    CoWrReset = 0x02,
    CoRdVersion = 0x03,
    CoRdSysLog = 0x04,
    CoWrIdbase = 0x07,
    CoRdIdbase = 0x08,
    CoWrLearnmode = 0x17,
    CoRdLearnmode = 0x18,
}

/// RORG (radio choice byte) — only the v1 set we decode.
pub mod rorg {
    pub const RPS: u8 = 0xF6;      // §EEP F6
    pub const ONE_BS: u8 = 0xD5;   // §EEP D5
    pub const FOUR_BS: u8 = 0xA5;  // §EEP A5
    pub const VLD: u8 = 0xD2;      // §EEP D2
    pub const UTE: u8 = 0xD4;      // §EEP D4
    pub const ADT: u8 = 0xA6;      // §EEP A6 wrapper
    pub const SIGNAL: u8 = 0xD0;
    pub const MSC: u8 = 0xD1;
}
